swarm_transport.go

```diff
@@ -26,8 +27,31 @@ func (s *Swarm) TransportForAddr(a ma.Multiaddr) transport.Transport {
 			return transport
 		}
 	}
-	p := protocols[len(protocols)-1]
-	return s.transports.m[p.Code]
+
+	return s.transports.m[protocols[len(protocols)-1].Code]
+}
+
+// TransportForListening retrieves the appropriate transport for listening on
+// the given multiaddr.
+func (s *Swarm) TransportForListening(a ma.Multiaddr) transport.Transport {
+	protocols := a.Protocols()
+	if len(protocols) == 0 {
+		return nil
+	}
+
+	s.transports.RLock()
+	defer s.transports.RUnlock()
+	selected := s.transports.m[protocols[len(protocols)-1].Code]
```

Comment:
will this work for TCP? don't we have `P_TCP` followed by `P_IP` there?

Reply:
Other way around. We have IP followed by TCP. The idea is that, except for proxies like p2p-circuit, each multiaddr protocol "wraps" the previous one. So we pick the last one and then let it deal with the previous ones recursively (possibly delegating back to the swarm or some other transport manager). I'd rather build up from left to right but that starts to be come an abstraction nightmare.

> will

FYI, all this code *does* actually work in practice. I've been running go-ipfs with these changes (all the refactor code) for a while with no issue.

Reply:
Yeah, it looks good overall; just getting a little itchy with the possibility of introducing some very subtle bug.

Reply:
I agree. And I know these are a *lot* of changes. That's why I'm letting them bake all together for a while while they're reviewed instead of merging them piece-wise.